language_service/src/hover.rs

```diff
@@ -188,18 +193,38 @@ impl Visitor<'_> for HoverVisitor<'_> {
 }
 
 fn markdown_with_doc(doc: &Rc<str>, code: impl Display) -> String {
-    if doc.is_empty() {
+    let parsed_doc = parse_doc(doc);
+    if parsed_doc.summary.is_empty() {
         markdown_fenced_block(code)
     } else {
         format!(
             "{}
 {}",
-            doc,
+            parsed_doc.summary,
             markdown_fenced_block(code)
         )
     }
 }
 
+fn parse_doc(doc: &str) -> Documentation {
+    let summary_re =
+        Regex::new(r"(^|(\r?\n))\s*#\s*((S|s)ummary)[\s\r\n]*").expect("Invalid regex");
+    let header_re = Regex::new(r"\r?\n\s*#\s*(\w+)[\s\r\n]*").expect("Invalid regex");
```

Comment:
The word character (`\w`) doesn't match whitespace, so I don't think this would match something like `# Some header`. It also wouldn't match any punctuation if used (e.g. the single quote in `# Here's stuff` would break it)

Basically, you just want a line to start with `# ` and then contain anything to end of line. Cleaning up similar to the above, something like the below should do it.

`(?m)^# .+$`

Reply:
I've used a variation of the regex you proposed below.